### ISSUE TITLE
lzcnt, tzcnt, popcnt false depndency not handled in CoreCLR

### ISSUE THREAD

**Issue by damageboy** [dotnet/coreclr#19555]
Hi,
Following my successful foray into the brave world of CoreCLR architecture specific intrinsics I think I stumbled upon a relatively known bug/deficiency in most intel CPUs (I have no knowledge on how this affects AMD CPUs if at all) that is now affecting perf on CoreCLR.

It appears that when I write a tight / unrolled loop with any of the intrinsics mentioned in the issue name:
* `lzcnt` (`LeadingZeroCount()`)
* `tzcnt` (`TrailingZeroCount()`)
* last but certainly not least popcnt (`popcnt` (`PopCount()`)

CoreCLR generates seemingly very good code when inspecting the JIT output, however that code hits a rather known Intel false dependency bug, that has been covered quite extensively:
*  [GCC mailing list](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=62011)
* [This disturbingly detailed explanation and insight into the underlying implementation](https://stackoverflow.com/questions/21390165/why-does-breaking-the-output-dependency-of-lzcnt-matter)
* [Dan Luu's post about popcnt](https://danluu.com/assembly-intrinsics/) (Near the end of the post)

The gist of the matter is that most in-production Intel CPUs, from Sandy-Bridge and onward up to, and including Skylake (for `popcnt`) and not-including Skylake (for `lzcnt`, `tzcnt`) introduce a dependency on the DESTINATION register for those instructions.

So a loop, written in c# as:
```csharp
for (var i = 0; i < n; i+=4, bits += 4)
    index += PopCount(bits[0]) + PopCount(bits[1]) + PopCount(bits[2]) + PopCount(bits[3]);}
```
Get's compiled/jitted down to (copy-pasted from CoreCLR JITDump):
```asm
G_M2736_IG05:        ; func=00, offs=000069H, size=0030H, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, byref, isz
IN001e: 000069 F34C0FB816           popcnt   r10, qword ptr [rsi]
IN001f: 00006E 4903D2               add      rdx, r10
IN0020: 000071 F34C0FB85608         popcnt   r10, qword ptr [rsi+8]
IN0021: 000077 4903D2               add      rdx, r10
IN0022: 00007A F34C0FB85610         popcnt   r10, qword ptr [rsi+16]
IN0023: 000080 4903D2               add      rdx, r10
IN0024: 000083 F34C0FB85618         popcnt   r10, qword ptr [rsi+24]
IN0025: 000089 4903D2               add      rdx, r10
Hope I've done a half-decent job in tracking this....
IN0026: 00008C 4183C104             add      r9d, 4
IN0027: 000090 4883C620             add      rsi, 32
IN0028: 000094 453BC8               cmp      r9d, r8d
IN0029: 000097 7CD0                 jl       SHORT G_M2736_IG05
```

As you can easily read/see, each `popcnt` instruction in this unrolled loop (and also between loop iterations) is writing its output into `r10`, thus, according to the described errata/cpu bug, is causing the next `popcnt` to stall before it can proceed with actual execution.

Given that the CPU can, in theory, execute two such `popcnt` instruction in parallel, this false dependency seems to have been messing around with various people's code perf for a long time.

I think I even found an inadvertent "complaint" about this inside this [CoreCLR issue](https://github.com/dotnet/coreclr/issues/15506#issuecomment-353674452) where @Tornhoof and @fiigii  are trying to make a popcnt loop / unrolled loop perform faster but aren't achieving substantial speedup.

It seems like modern compilers, such as GCC, can and do alleviate the impact of this issue by [clearing the bottom 32 bits of each destination preemptively before emitting the `popcnt`instruction](https://godbolt.org/z/pSr6iu), copy pasted to here for completeness...:

```asm
.L3:
        xor     edx, edx
        xor     ecx, ecx
        popcnt  rcx, QWORD PTR [rdi]
        popcnt  rdx, QWORD PTR [rdi+8]
        add     rdx, rcx
        xor     ecx, ecx
        popcnt  rcx, QWORD PTR [rdi+16]
        add     rdx, rcx
        xor     ecx, ecx
        popcnt  rcx, QWORD PTR [rdi+24]
        add     rdi, 32
        add     rdx, rcx
        add     eax, edx
        cmp     rdi, rsi
        jne     .L3
```

Would this be something that CoreCLR JIT could dynamically detect (according to CPU model/family) and insert into the instruction stream?

**Comment by AndyAyersMS** [comment]
@fiigii -- is this something you'd recommend fixing?
cc @dotnet/jit-contrib

**Comment by damageboy** [comment]
For what it's worth, from my C++ benchmarking, I can testify that unrolling a loop of 4 POPCNTs without this `xor dest,test` hack (I used inline assembly to force the compiler generate the suboptimal code) led to a 80% regression for 4096 bit arrays, for example...

In other words the gcc generate `POPCNT` ran at `X` ns for 4096 bits, while the unoptimized `POPCNT` ran at 1.8 `X`...

I'm writing a blog post about what I'm doing, and I can incorporate those tests if anyone is interested...?

**Comment by GSPP** [comment]
Isn't `xor dst, dst` done as part of register renaming on modern CPUs? It might cost exactly zero latency. Maybe it is the best strategy to just always insert it.

**Comment by fiigii** [comment]
> Isn't xor dst, dst done as part of register renaming on modern CPUs?

Yes, but CPU front-end resource is also limited. Too much XOR will impact the front-end performance by renaming and code size.

> Maybe it is the best strategy to just always insert it.

We may need more scenarios and data to decide the implementation approach.

**Comment by damageboy** [comment]
@fiigii I think that what @GSPP meant was that it might simply be a good strategy to always insert it for **those** 3 affected instructions *on the relevant* CPU models..

A simple [experiment with gcc](https://godbolt.org/z/6K2D1q) seems to indicate that @GSPP is right:
gcc seems to just blindly insert a `xor dest,dest` whenever it sees a `POPCNT` on skylake (in that example), even if there isn't a loop involved (see example).

I completely understand what you are trying to say about the CPU front-end being a limited resource in the general sense, and that over-utililzing it also needs to be taken into consideration, it's just that for these specific instructions the benefit out-weighs the cost...

I don't think I've ever seen a `POPCNT` done outside of a loop in my life :), and the gcc behavior around a single `POPCNT` seems to imply that they haven't either :)

In addition, given that many devs deliberately take advantage of Intel CPUs ability to handle 2 independent `POPCNT` instructions in parallel on two different execution ports (If memory serves me well), missing the opportunity to free the CPU from stalling carries a heavy penalty given the context.

As such, **blindly** clearing the destination in order to trigger the register renaming is probably the sane thing to do for this bunch of orders, while definitely NOT the right thing to do in other cases...

Do you have any suggestions on how else we can collect data / scenarios to be able to come to a decision...?

**Comment by fiigii** [comment]
@damageboy Thank you so much for the investigation and clarification. Looks like we can give a try with unconditionally inserting XOR.

However, before fixing the issue in codegen, we need more infrastructure work in the runtime for the micro-architecture specific optimization. Currently, CoreCLR is only aware of the underlying hardware difference at ISA levels, so we need to make some change to let RyuJIT know the micro-architecture. After that, we can only insert XOR for `lzcnt`/`tzcnt` on SNB/IVB and HSW, and remove XOR for `popcnt` in the future.

@AndyAyersMS @CarolEidt @jkotas Do you agree to add the micro-architecture checks?
